### ISSUE TITLE
Factory building manu speed is always 1.0, MaxProd/MaxConsumed follows slugs amount inserted

### ISSUE THREAD

**Issue by featheredtoast** [porisius/FicsitRemoteMonitoring#29]
Similar to the miners bug, factory buildings are also getting MaxProd numbers from the # of slugs inserted rather than the set clockspeed.

Also, ManuSpeed always seems to return as `1.0`. Probably desirable to follow the same as extractors where 100 (for 100%) to 250 (for 250%) etc.

Example Alclad Aluminum Sheet assembler, set to 222% clockspeed. (expected 66.6 aluminum ingots 22.2 copper ingots in, 66.6 aluminum sheets out)

```
{
      "building":"Assembler",
      "location":{
         "x":15259.248047,"y":46446.4375,"z":25134.820312,
         "rotation":130
      },
      "Recipe":"Alclad Aluminum Sheet",
      "production":[
         {
            "Name":"Alclad Aluminum Sheet",
            "CurrentProd":0.0,
            "MaxProd":75.0,
            "ProdPercent":0.0
         }
      ],
      "ingredients":[
         {
            "Name":"Aluminum Ingot",
            "CurrentConsumed":0.0,
            "MaxConsumed":75.0,
            "ConsPercent":0.0
         },         {
            "Name":"Copper Ingot",
            "CurrentConsumed":0.0,
            "MaxConsumed":25.0,
            "ConsPercent":0.0
         }
      ],
      "ManuSpeed":1.0,
      "IsConfigured":true,
      "IsProducing":false,
      "IsPaused":false,
      "CircuitID": 0
}
```

**Comment by porisius** [comment]
Unfortunately, I have not had the mental strength due to being sick to test this properly, but if you could give it a spot, please. I copied what I did to fix the getExtractor.
[FicsitRemoteMonitoring.zip](https://github.com/porisius/FicsitRemoteMonitoring/files/10357386/FicsitRemoteMonitoring.zip)

**Comment by featheredtoast** [comment]
I took this patch for a spin -- I've observed that getFactory entries are 1/10th of their expected production+consumption speeds. (I also had to start producing before ManuSpeed gets updated, but I'm assuming that's how that particular function goes... otherwise ManuSpeed looks much better!)

Expected production to be at 65, consumption to be 65+22 (21.667) respectively.

```
{
      "building":"Assembler",
      "location":{
         "x":-51167.96875,"y":256804.21875,"z":-2780.661133,
         "rotation":90
      },
      "Recipe":"Alclad Aluminum Sheet",
      "production":[
         {
            "Name":"Alclad Aluminum Sheet",
            "CurrentProd":6.5,
            "MaxProd":6.5,
            "ProdPercent":99.599998
         }
      ],
      "ingredients":[
         {
            "Name":"Aluminum Ingot",
            "CurrentConsumed":6.5,
            "MaxConsumed":6.5,
            "ConsPercent":99.599998
         },         {
            "Name":"Copper Ingot",
            "CurrentConsumed":2.2,
            "MaxConsumed":2.2,
            "ConsPercent":99.599998
         }
      ],
      "ManuSpeed":216.699997,
      "IsConfigured":true,
      "IsProducing":true,
      "IsPaused":false,
      "CircuitID": 29
}
```

getProdStats entry is a little different- CurrentProd is similarly 1/10th expected, but MaxProd is 7.0... assuming this is due to rounding happening somewhere. (May I also suggest keeping rounding in all places to 3 decimal places, to sync with in game machine UI.)

Expected CurrentProd and MaxProd to both be 65 here.

```
{
      "ItemName":"Alclad Aluminum Sheet",
      "ProdPerMin":"P: 6.0/7.0/min - C: 0.0/0.0/min",
      "ProdPercent":99.610001,
      "ConsPercent":0.0,
      "CurrentProd":6.0,
      "MaxProd":7.0,
      "CurrentConsumed":0.0,
      "MaxConsumed":0.0,
      "Type":"Belt"
}
```

Don't stress too much on this before you're recovered!

**Comment by porisius** [comment]
Odd... what does fluids look like? I wonder if they are just as off.

**Comment by featheredtoast** [comment]
Ah, new theory -- production rates are looking only looking at items per production cycle only, and not multiplying by production cycles per minute.

Threw it at a refinery today at wet concrete, and the liquids follow similar numbers:

concrete prod is 7 (expected 140, max 140)
Limestone consumption is 10.5, max 15 (expected 210, max 210)
Water is 8.7/8.8 (expected 175, max 175)

The theory about it coming from production cycles without muliplying by cycles per minute is from inspecting the recipe:
4 concrete (4 * 1.75 = 7)
Limestone: 6 (6 * 1.75 = 10.5)
Water: 5 (5 * 1.75 = 8.75 rounding up is 8.8)

```
{
      "building":"Refinery",
      "location":{
         "x":-84420.929688,"y":233475.9375,"z":-2160.155762,
         "rotation":-20
      },
      "Recipe":"Alternate: Wet Concrete",
      "production":[
         {
            "Name":"Concrete",
            "CurrentProd":7.0,
            "MaxProd":7.0,
            "ProdPercent":99.599998
         }
      ],
      "ingredients":[
         {
            "Name":"Limestone",
            "CurrentConsumed":10.5,
            "MaxConsumed":10.5,
            "ConsPercent":99.599998
         },         {
            "Name":"Water",
            "CurrentConsumed":8.7,
            "MaxConsumed":8.8,
            "ConsPercent":99.599998
         }
      ],
      "ManuSpeed":175.0,
      "IsConfigured":true,
      "IsProducing":true,
      "IsPaused":false,
      "CircuitID": 37
}
```

**Comment by porisius** [comment]
Don't worry... appears I am grabbing from the recipe's required items instead of the items per minute...
![image](https://user-images.githubusercontent.com/47161774/211105710-6658e11a-33f2-4b59-86dd-3ecc1f1f8c6b.png)

I'll take a look at this more later though. The idea of debug commands to get the raw info has been a blessing.

**Comment by porisius** [comment]
I'm struggling on how to solve this... the output data won't change much as it is easy to determine productivity and efficiency, but input... there isn't a way to accurately determine rate/efficiency that I can think of... however, I can get the inventory of each component item as well as the required amount for production.

I wish I had access to someone who used to work in a factory, but alas cancer took my father from me. I might still have contact with a quality director that has worked in the automotive industry. I will see if I can reach out to him for insight. Obviously, I'll take anything, but I am looking at/for sources of insight.

**Comment by featheredtoast** [comment]
Wow, my condolences, I'm so sorry for your loss. That's tough.

I don't have familiarity with the tools or functions available, my assumption was there would be a "cycles per minute" function for the machine, which could be used to multiply against the existing numbers for both production and consumption, but it sounds like that's less available than I had thought.

**Comment by featheredtoast** [comment]
Would something like `getProductionCycleTime` work? (or, I think it's CalcProdCycleTime in your debug already?)

![image](https://user-images.githubusercontent.com/1322534/212461450-27d7ef0d-72d6-40c9-b4cf-921a87f81c72.png)

If so, (recipe) cycle time can be used to calculate cycles per minute (I'm assuming it returns the number in seconds):
`recipe cycles per minute = 60 (seconds)/(recipe cycle time)`

You also have productivity and currentPotential (which I think is clockspeeds, between 0.1 and 2.5?) -- which can be used as a straight multiplier on the base recipe's clockcycle.
Then:
`recipe cycles per minute * recipe ingredient amount * productivity` = consumption per minute
`recipe cycles per minute * recipe product amount * productivity` = production per minute
`recipe cycles per minute * recipe ingredient amount * currentPotential` = max consumption per minute
`recipe cycles per minute * recipe product amount * currentPotential` = max production per minute

I am working off of names only, so this might be completely off base. That'd be my best guess as to how to calculate /minute rates.

Like always, though, please don't stress about any of this until you're ready to, especially now given the tough time you and your family is going through.

**Comment by porisius** [comment]
The old man past away back in 2017, so no, not that rough... Just wish I had his wisdom.

I might look into that, game convention in town this weekend, so I am going to take a break and clear my head there.

**Comment by porisius** [comment]
Okay, I sent FRM 0.8.17 to SMR. I would have gotten done sooner, but other projects popped up that I jumped to help move along.

**Comment by featheredtoast** [comment]
No problem- This looks great, thank you!

**Comment by featheredtoast** [comment]
Spoke too soon -- Let me rephrase, new numbers look good for getExtractor and getFactory (yay!)

However, numbers on getProdStats may need revisiting. I'm testing with wet concrete again - the building itself has aligned numbers through getExtractor and getFactory, but the getProdStats aren't matching.

A test refinery production wet concrete, producing 53, with 142 max production:
```
{
      "building":"Refinery",
      "location":{
         "x":-86558.179688,"y":230281.609375,"z":-2160.15625,
         "rotation":90
      },
      "Recipe":"Alternate: Wet Concrete",
      "production":[
         {
            "Name":"Concrete",
            "CurrentProd":53.599998,
            "MaxProd":142.399994,
            "ProdPercent":37.700001,
            "Inventory":0
         }
      ],
      "ingredients":[
         {
            "Name":"Limestone",
            "CurrentConsumed":80.5,
            "MaxConsumed":213.600006,
            "ConsPercent":37.700001,
            "Inventory":100
         },         {
            "Name":"Water",
            "CurrentConsumed":67.099998,
            "MaxConsumed":178.0,
            "ConsPercent":37.700001,
            "Inventory":11097
         }
      ],
      "ManuSpeed":178.0,
      "IsConfigured":true,
      "IsProducing":true,
      "IsPaused":false,
      "CircuitID": 28
}
```

getProdStats' entry for concrete looks a little low:
```
{
      "ItemName":"Concrete",
      "ProdPerMin":"P: 15.0/20.0/min - C: 0.0/0.0/min",
      "ProdPercent":76.82,
      "ConsPercent":0.0,
      "CurrentProd":15.0,
      "MaxProd":20.0,
      "CurrentConsumed":0.0,
      "MaxConsumed":0.0,
      "Type":"Belt"
}
```

I can close out this one and add a new issue for this one if you'd like.

**Comment by porisius** [comment]
Nah, technically the same issue... thing is, we got getFactory working, and the only real difference between getFactory and getProdStats is one is by FGBuildableManufacturer (getFactory) and the other is by FGItemDescriptor (getProdStats). So, I ought to be able to copy one to another and have both work.

**Comment by porisius** [comment]
well, I copied to getProdStats... but... SMR is having issues and won't let me upload. I notified the SMR Infra-Team about it.

**Comment by porisius** [comment]
0.8.18 is live, please confirm on your end.

**Comment by featheredtoast** [comment]
Hmm, now in getProdStats, non-raw materials are seen as "invalid" now.

I'm building Control Rods, and Concrete, but neither of them are showing up in the response - note the N/A item name, with Invalid type:

```
[{
      "ItemName":"Coal",
      "ProdPerMin":"P:4.0/min - C: 0.0/min",
      "ProdPercent":0.78,
      "ConsPercent":0.0,
      "CurrentProd":4.0,
      "MaxProd":480.0,
      "CurrentConsumed":0.0,
      "MaxConsumed":0.0,
      "Type":"Solid"
}, {
      "ItemName":"Copper Ore",
      "ProdPerMin":"P:600.0/min - C: 0.0/min",
      "ProdPercent":100.0,
      "ConsPercent":0.0,
      "CurrentProd":600.0,
      "MaxProd":600.0,
      "CurrentConsumed":0.0,
      "MaxConsumed":0.0,
      "Type":"Solid"
}, {
      "ItemName":"Iron Ore",
      "ProdPerMin":"P:240.0/min - C: 0.0/min",
      "ProdPercent":100.0,
      "ConsPercent":0.0,
      "CurrentProd":240.0,
      "MaxProd":240.0,
      "CurrentConsumed":0.0,
      "MaxConsumed":0.0,
      "Type":"Solid"
}, {
      "ItemName":"Limestone",
      "ProdPerMin":"P:505.0/min - C: 0.0/min",
      "ProdPercent":81.139999,
      "ConsPercent":0.0,
      "CurrentProd":505.0,
      "MaxProd":623.0,
      "CurrentConsumed":0.0,
      "MaxConsumed":0.0,
      "Type":"Solid"
}, {
      "ItemName":"N/A",
      "ProdPerMin":"P:8.0/min - C: 4.0/min",
      "ProdPercent":99.610001,
      "ConsPercent":99.610001,
      "CurrentProd":8.0,
      "MaxProd":8.0,
      "CurrentConsumed":4.0,
      "MaxConsumed":4.0,
      "Type":"Invalid"
}, {
      "ItemName":"Water",
      "ProdPerMin":"P:348.0/min - C: 0.0/min",
      "ProdPercent":79.040001,
      "ConsPercent":0.0,
      "CurrentProd":348.0,
      "MaxProd":440.0,
      "CurrentConsumed":0.0,
      "MaxConsumed":0.0,
      "Type":"Liquid"
}]
```

**Comment by porisius** [comment]
Wait what? What does getFactory say about it then?

**Comment by porisius** [comment]
The Type change, I know about... I took the ENUM from the item...

**Comment by featheredtoast** [comment]
getFactory still looks good from what I could tell. :+1:

**Comment by porisius** [comment]
but... it's a copy of... I'll look into it in a minute... at least SMR Uploads is back online

**Comment by featheredtoast** [comment]
Looks like the last patch fixed up the production numbers - thanks for your work on this! :tada:
